### PR TITLE
Implement cf deployment

### PR DESCRIPTION
**Pull request by joshq00**

### CODE REVIEW

File: checksum.sh [hunk 1]
```diff
@@ -0,0 +1 @@
+drone secure --repo josh/drone-cloudfoundry --checksum
```
Comment (tboerger) [inline checksum.sh]: That's nothing for the official repo.

File: .drone.yml [hunk 2]
```diff
@@ -1,9 +1,16 @@
...
+  environment:
+    - GO15VENDOREXPERIMENT=0
+    - GOOS=linux
+    - GOARCH=amd64
+    - CGO_ENABLED=0
   commands:
-    - make deps
-    - make build
-    - make test
+    - go get
```
Comment (tboerger) [inline .drone.yml]: Please keep the Malefile and the make tasks for the .drone.yml

File: DOCS.md [hunk 3]
```diff
@@ -1,11 +1,42 @@
...
+    api: api.run.pivotal.io
+    user: $$USER
+    password: $$PASSWORD
+    org: $$USER
+    space: production
+    when:
+      branch: master
 ```
+
+## Available options
```
Comment (tboerger) [inline DOCS.md]: I would prefer the same documentation as for the other plugins with the removed list above. Beside that please use only underscores and no dashes for the config variables.

File: Dockerfile [hunk 4]
```diff
@@ -5,10 +5,7 @@
 
 FROM alpine:3.2
 
-RUN apk update && \
```
Comment (tboerger) [inline Dockerfile:8]: Please keep it as multiline.

File: README.md [hunk 5]
```diff
@@ -5,6 +5,8 @@
 
 Drone plugin for deploying to Cloud Foundry
 
+See [DOCS.md](DOCS.md) for `.drone.yml` usage
```
Comment (tboerger) [inline README.md]: I would drop that.

File: commands.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	org, space := vargs.Org, vargs.Space
+	require("org", org)
+	require("space", space)
+	fmt.Printf("Targeting %s:%s...\n", org, space)
+	return exec.Command("cf", "target", "-o", org, "-s", space)
+}
+
+// cf deploy
+func deploy(workspace drone.Workspace, app App, route Route, flags Flags) *exec.Cmd {
+	fmt.Println("Deploy")
```
Comment (tboerger) [inline commands.go]: Drop that debug output.

File: main.go [hunk 7]
```diff
@@ -1,14 +1,22 @@
 package main
 
 import (
-	"fmt"
-)
-
-var (
-	build     string
```
Comment (tboerger) [inline main.go:8]: Please keep the build and the buildDate, that's getting set within the makefile

File: main.go [hunk 8]
```diff
@@ -1,14 +1,22 @@
...
-	buildDate string
+	"github.com/drone/drone-go/drone"
+	"github.com/drone/drone-go/plugin"
 )
 
 func main() {
-	fmt.Printf("Drone Cloud Foundry Plugin built at %s\n", buildDate)
```
Comment (tboerger) [inline main.go:13]: Please keep that also.

File: util.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	io.Copy(os.Stderr, stderr)
+	err := cmd.Wait()
+
+	if err != nil {
+		os.Exit(1)
+	}
+}
+
+// Require that a field has been set
+func require(field string, val interface{}) {
```
Comment (tboerger) [inline util.go:34]: That's useful for nearly all plugins, that should be moved to drone-go repo.

File: README.md (hunk 5 above)
Comment (joshq00) [inline README.md]: I would suggest adding this to all the other plugins. The difference between _how to use this in development_ and _how to use this in practice_ should both be readily available in the readme. I actually think _how to use this in practice_ is more important and should be the one in `README.md`. Whereas `DOCS` would make more sense for development

Comment (tboerger) [inline README.md]: The main entry point for the plugins is the plugin index, on the index you always see DOCS.md for every plugin and no README.md

Comment (joshq00) [inline README.md]: Fair enough

Comment (tboerger) [inline README.md]: But I don't have a strong opinion on that single part if it gets added to all official plugins.

**Comment by joshq00** [comment]
Any further issues, @tboerger?

File: .drone.yml [hunk 10]
```diff
@@ -1,5 +1,10 @@
 build:
   image: golang:1.5
+  environment:
+    - GO15VENDOREXPERIMENT=0
+    - GOOS=linux
+    - GOARCH=amd64
+    - CGO_ENABLED=0
```
Comment (tboerger) [inline .drone.yml]: No need for those env variables :)

File: Makefile [hunk 11]
```diff
@@ -20,6 +20,9 @@ test:
 
 build:
 	go build -ldflags '-s -w $(LDFLAGS)'
+	go get github.com/cloudfoundry/cli/main
+	go get github.com/cloudfoundry/cli/main
```
Comment (tboerger) [inline Makefile]: Duplicate of the line above and move the get to the deps task

File: main.go [hunk 12]
```diff
@@ -11,4 +14,18 @@ var (
 
 func main() {
 	fmt.Printf("Drone Cloud Foundry Plugin built at %s\n", buildDate)
+
+	cfargs := CloudFoundry{}
+	workspace := drone.Workspace{}
+	plugin.Param("workspace", &workspace)
+	plugin.Param("vargs", &cfargs)
+	plugin.Parse()
```
Comment (tboerger) [inline main.go]: MustParse

**Comment by tboerger** [comment]
Some tiny changes, adterwards LGTM

File: main.go (hunk 12 above)
Comment (joshq00) [inline main.go]: ~~Can you elaborate?~~
Ok. Will use `plugin.MustParse()`

File: Makefile (hunk 11 above)
Comment (joshq00) [inline Makefile]: Whoopsie. Will do

**Comment by tboerger** [comment]
I will add another followup to update the DOCS.md a little bit, but otherwise LGTM

**Comment by joshq00** [comment]
@tboerger : will this be available soon? I saw the build was missing credentials to publish to docker hub

Right now, we're using a custom `PLUGIN_FILTER`, and I'd like to get away from that

**Comment by tboerger** [comment]
Brad needs to add the secrets.

**Pull request merged**